src/adapter/src/catalog/migrate.rs

```diff
@@ -182,6 +209,567 @@ pub(crate) async fn migrate(
+                    | CreateSourceConnection::LoadGenerator {
+                        generator:
+                            LoadGenerator::Clock
+                            | LoadGenerator::Datums
+                            | LoadGenerator::Counter
+                            | LoadGenerator::KeyValue,
+                        ..
+                    }),
+                name,
+                col_names,
+                include_metadata,
+                format,
+                envelope,
+                mut with_options,
+                if_not_exists,
+                in_cluster,
+                progress_subsource,
+                external_references,
+                key_constraint,
+            }) => {
+                // To check if this is a source that has already been migrated we use a basic
+                // heuristic: if there is at least one existing table for the source, and if
+                // the envelope/format/include_metadata options are empty, we assume it's
+                // already been migrated.
+                let tables_for_source =
+                    items_with_statements_copied
+                        .iter()
+                        .any(|(item, statement)| match statement {
+                            Statement::CreateTableFromSource(stmt) => {
+                                let source: GlobalId = match &stmt.source {
+                                    RawItemName::Name(_) => {
+                                        unreachable!("tables store source as ID")
+                                    }
+                                    RawItemName::Id(source_id, _, _) => {
+                                        source_id.parse().expect("valid id")
+                                    }
+                                };
+                                source == item.id
+                            }
+                            _ => false,
+                        });
+                if tables_for_source
+                    && envelope.is_none()
+                    && format.is_none()
+                    && include_metadata.is_empty()
+                {
+                    info!("migrate: skipping already migrated source: {}", name);
+                    return Ok(());
+                }
+
+                // Use the current source name as the new table name, and rename the source to
+                // `<source_name>_source`. This is intended to allow users to continue using
+                // queries that reference the source name, since they will now need to query the
+                // table instead.
+                let new_source_name_ident = Ident::new_unchecked(
+                    name.0.last().expect("at least one ident").to_string() + "_source",
+                );
+                let mut new_source_name = name.clone();
+                *new_source_name.0.last_mut().expect("at least one ident") = new_source_name_ident;
```

Comment:
This would fail if there already happens to be an object with the name `<source_name>_source` right? I think we need to check that this name is unique, and maybe try appending a different number until it becomes unique.

Reply:
ah good point - will update